packages/ai-ide/src/browser/template-preference-contribution.ts

```diff
@@ -0,0 +1,118 @@
+// *****************************************************************************
+// Copyright (C) 2024 EclipseSource GmbH.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { FrontendApplicationContribution, PreferenceService } from '@theia/core/lib/browser';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { FrontendPromptCustomizationServiceImpl } from '@theia/ai-core/lib/browser/frontend-prompt-customization-service';
+import {
+    PROMPT_TEMPLATE_WORKSPACE_DIRECTORIES_PREF,
+    PROMPT_TEMPLATE_WORKSPACE_EXTENSIONS_PREF,
+    PROMPT_TEMPLATE_WORKSPACE_FILES_PREF
+} from './workspace-preferences';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { Path } from '@theia/core';
+
+@injectable()
+export class TemplatePreferenceContribution implements FrontendApplicationContribution {
+
+    @inject(PreferenceService)
+    protected readonly preferenceService: PreferenceService;
+
+    @inject(FrontendPromptCustomizationServiceImpl)
+    protected readonly customizationService: FrontendPromptCustomizationServiceImpl;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    onStart(): void {
+        Promise.all([this.preferenceService.ready, this.workspaceService.ready]).then(() => {
+            // Set initial template directories, extensions, and files from preferences
+            this.updateTemplateDirectories();
+            this.updateTemplateFileExtensions();
+            this.updateTemplateFiles();
```

Comment:
All three calls will each lead to a full rescan of all files, disposing and recreation of file watchers etc. We should think about an (re)-initialize call in the service to update all 3 data points (actually 4, including the preference which is currently managed by `ai-core-preferences`) at the same time to avoid this. Same for the workspace root changes below.

Reply:
This makes sense. Basically I would add an initialize method to the frontend-prompt-customization-service where I can hand in all three parameters at once and only update once after this.
Does this make sense to you?

As the ai-core-preference comes from a different contribution, I am not sure, whether we can avoid this duplication, though?

Reply:
You're right, I think it becomes too clunky to avoid the duplication there.

But also it would be weird to have a `initialize` method which only takes 3 of the 4 parameters.

Suggestion: Declare a single `update` method in the service which takes an object as a parameter, with an optional entry for the properties to update. Then it can set all of the handed over parameters at once, performing the internal update also once. This makes the service interface nicer.

Of course we still have the duplicated effort then, with the update once called with one parameter and then again with 3 parameters, but I think we can live with that for now.

Reply:
I have now introduced a update function for the three external preferences. Did you mean to also include the 4th (internal) preference? If so, we should extract this preference from the service and set it from a frontendApplicationContribution in ai-core. Please quickly confirm if you like this or if we should keep the preference for the global directory as is (for now) @sdirix